ecs-agent/netlib/data/network_client.go

```diff
@@ -94,3 +132,127 @@ func (ndc *networkDataClient) GetNetworkNamespace(netNSName string) (*tasknetwor
 
 	return netNS, nil
 }
+
+// AssignGeneveDstPort returns an unused destination port number for GENEVE interfaces.
+// By default for a particular VNI, it will return the default GENEVE destination port - 6081.
+// In case port 6081 is taken by another interface using the same VNI, it will chose a
+// random port from within the pre-configured range.
+// Once a port number is assigned by this function, it won't be used until
+// ReleaseGeneveDstPort is called with the port and VNI.
+func (ndc *networkDataClient) AssignGeneveDstPort(vni string) (uint16, error) {
+	metricEntry := ndc.metricsFactory.New(metrics.AssignGeneveDstPortMetricName)
+	port, err := ndc.assignPort(geneveDstPortMin, geneveDstPortMax, vni, geneveDstPortDefault)
+	metricEntry.Done(err)
+
+	if errNoValue == errors.Cause(err) {
+		ndc.metricsFactory.New(metrics.V2NDestinationPortExhaustedMetricName).Done(err)
+	}
+	return port, err
+}
+
+func (ndc *networkDataClient) assignPort(min, max int, vni string, defaultValue int) (uint16, error) {
+	var value int
+
+	err := ndc.DB.Update(func(tx *bolt.Tx) error {
+		var e error
+
+		// Destination ports are required to be unique only per VNI. Hence we create nested buckets
+		// inside the port bucket with the VNI as the name.
+		portBucket := tx.Bucket([]byte(portBucketName))
+		if vni != "" {
+			if portBucket, e = portBucket.CreateBucketIfNotExists([]byte(vni)); e != nil {
+				return e
+			}
+		}
+
+		value, e = findRandomUnusedValue(portBucket, min, max, defaultValue)
+		return e
+	})
+
+	return uint16(value), err
+}
+
+func (ndc *networkDataClient) releasePort(port uint16, namespace string) error {
+	return ndc.DB.Update(func(tx *bolt.Tx) error {
+		var e error
+		portBucket := tx.Bucket([]byte(portBucketName))
+		if namespace != "" {
+			if portBucket, e = portBucket.CreateBucketIfNotExists([]byte(namespace)); e != nil {
+				return e
+			}
+		}
+
+		key := []byte(strconv.Itoa(int(port)))
+		return portBucket.Delete(key)
+	})
+}
+
+func (ndc *networkDataClient) ReleaseGeneveDstPort(port uint16, vni string) error {
+	metricEntry := ndc.metricsFactory.New(metrics.ReleaseGeneveDstPortMetricName)
+	err := ndc.releasePort(port, vni)
+	metricEntry.Done(err)
+	return err
+}
+
+// findRandomUnusedValue returns a random integer value that was previously unused.
+// You may specify a defaultValue if you have a preferred value. If the defaultValue is already used,
+// it will return a random unused value from within the range. An example for when we use the defaultValue
+// is while assigning destination ports for GENEVE interfaces. By default we need to assign port 6081 unless
+// it is used by another existing interface. If 6081 is taken, we need to pick a random port from the specified
+// range of [min, max).
+// The function makes use of a data store bucket to store all the used values.
+func findRandomUnusedValue(bucket *bolt.Bucket, min, max, defaultValue int) (int, error) {
```

Comment:
I don't think any of the comments in this file mentioned whether this will be invoked concurrently. Is it possible for multiple clients to be invoking this method and accessing the same DB at the same time? If that's the case, should we add a semaphore to protect shared resources (in this case, pool of available port numbers)?

Reply:
I have removed implementation of NetworkDataClient from the shared library because the agent is able to inject the same into the network builder.